pkg/cloud/services/elb/loadbalancer.go

```diff
@@ -1604,6 +1528,164 @@ func (s *Service) reconcileV2LBTags(lb *infrav1.LoadBalancer, desiredTags map[st
 	return nil
 }
 
+// reconcileTargetGroupsAndListeners reconciles a Load Balancer's defined listeners with corresponding AWS Target Groups and Listeners.
+// These are combined into a single function since they are tightly integrated.
+func (s *Service) reconcileTargetGroupsAndListeners(spec *infrav1.LoadBalancer, lbSpec *infrav1.AWSLoadBalancerSpec) ([]*elbv2.TargetGroup, []*elbv2.Listener, error) {
+	existingTargetGroups, err := s.ELBV2Client.DescribeTargetGroups(
+		&elbv2.DescribeTargetGroupsInput{
+			LoadBalancerArn: aws.String(spec.ARN),
```

Comment:
For some reason, we're getting here without an ARN, which is causing failures.

Reply:
I think this is the reason:
```
diff --git a/pkg/cloud/services/elb/loadbalancer.go b/pkg/cloud/services/elb/loadbalancer.go
index 0e203e870..d9a5fe400 100644
--- a/pkg/cloud/services/elb/loadbalancer.go
+++ b/pkg/cloud/services/elb/loadbalancer.go
@@ -401,6 +401,7 @@ func (s *Service) createLB(spec *infrav1.LoadBalancer, lbSpec *infrav1.AWSLoadBa
        res := spec.DeepCopy()
        s.scope.Debug("applying load balancer DNS to result", "dns", *out.LoadBalancers[0].DNSName)
        res.DNSName = *out.LoadBalancers[0].DNSName
+       res.ARN = *out.LoadBalancers[0].LoadBalancerArn
        return res, nil
 }
```
Right after we create the LB, we don't set the ARN in the output. That only happens in `describeLB`.

Reply:
With the above change I got a successful install. But there is something I don't understand. In the first reconcile loop, the LB is created but we were not saving its ARN to the output so the TG creation failed. But in the subsequent reconcile loops, the LB exists and we should get the output from `describeLB` which should contain the ARN. It looks like the TG creation is not being attempted in that case. Maybe https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5004/files#diff-ee38bda5efe2606e7e7746a556c49a2d53d14eca60c029b8efebea657b8718f1R1551-R1553 is comparing `0 == 0` and returning `nil, nil, nil`.

Reply:
Answer: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5004/files#r1630996444

Reply:
Hmm, I wonder if I didn't move this assignment line over when relocating code? I'll review the diff, but thank you for going through this!

Reply:
Yes, that's it.